raui-core/src/application.rs

```diff
@@ -583,13 +588,83 @@ impl Application {
     /// [`process()`][Self::process] application, even if no changes have been detected
     #[inline]
     pub fn forced_process(&mut self) -> bool {
+        self.forced_process_with_context(&mut Default::default())
+    }
+
+    /// [`process()`][Self::process] application, even if no changes have been detected
+    #[inline]
+    pub fn forced_process_with_context<'a, 'b>(
+        &mut self,
+        process_context: &mut ProcessContext<'b>,
+    ) -> bool {
         self.dirty = true;
-        self.process()
+        self.process_with_context(process_context)
     }
 
     /// Process the application, updating animations, applying state changes, handling widget
     /// messages, etc.
+    #[inline]
     pub fn process(&mut self) -> bool {
+        self.process_with_context(&mut Default::default())
+    }
+
+    /// # Process Context
+    ///
+    /// The `process_context` argument allows you to optionally provide the UI's components with
+    /// mutable or immutable access to application data. This allows RAUI hosts to provide the UI
+    /// with direct access to application data, if necessary, instead of using
+    /// [`DataBinding`][crate::data_binding::DataBinding]s.
+    ///
+    /// This may also be used to facilitate different patterns of two-way communication between the
```

Comment:
Yeah, I was thinking about that afterwards and realized that it could easily be an anti-pattern to user certain patterns of communication with the host through the process context.

What I was immagining is that instead of being restricted to only using `Signal`s to communicate to the host from widgets, you could now put any kind of Rust channel sender or something similar in the `ProcessContext` and widgets could use that channel to send messages to the host. The host can also include multiple different channels for different things because it can add multiple different types to the process context.

Essentially it opens up practically _any_ kind of mechanism to talk between the host and the widgets, including my use case, such as giving the widgets a reference to the ECS world, allowing them to query from it and send events into the world.

Reply:
yeah, so my concerns were that right now communication follows strict rules: you can send messages to other widgets or you send signals to the host app, when we give users the ability for using other ways of communication, we might want to focus it to be only one way (widget -> app host) because if we allow the other way (app host -> widget), that won't be noticed by the application (right now it re-runs processing only when there is message sent to any widget or a change in widget state happen, and it should stay that way, otherwise i get PTSD from times when me and others had too much power with React and i restrict communication precisely to avoid that).

on the other hand i think users shouldn't be able to perform app host -> widget communication with this simply bc receivers would be required to be stored in widget state to keep them alive, and iirc regular receivers won't be allowed to do that without going into unsafe territory 🤔 

ok, i think it would be enough if we just replace the note about custom two-way communication with disclaimer that altho users are free to do it, they shouldn't because this might introduce dangerous and undefined behaviour on their side.

Reply:
> right now it re-runs processing only when there is message sent to any widget or a change in widget state happen, and it should stay that way

Yeah, I realized that in Bevy Retro and noticed that I was going to have to run `forced_process()` to make sure that the UI would get processed every frame.

Is that horribly un-performant? It makes it much easier to think about from a user perspective, at least in the context of Bevy where game systems typically run every frame, every time, and then they can think of UI similar to a system where it can access world data every frame and update it's contents if they desire.

It makes it more like an immediate mode GUI in the user-experience, which is nice, but I don't know if performance would become a problem with larger GUIs or something.

> ok, i think it would be enough if we just replace the note about custom two-way communication with disclaimer that altho users are free to do it, they shouldn't because this might introduce dangerous hidden bugs and undefined behaviour on their side.

OK, sounds good, I'll update that.

Reply:
well, ofc nothing stops you from enforcing re-run on every frame (using forced_process) but RAUI's goal has to be that by default tree (now whole tree, later only components that are actually changed or affected by other change) is rebuilt only when change occurs, the main goal is to make UI take as less processing time as possible (and until i complete props & states allocator and vdom diffing each process takes _some_ time that other game systems could take) ;p

Reply:
OK, I gave a solid break-down of why not to use a process context and I think I made it sound scarry enough. ;)